File: test/src/nn/modules/normalization/LayerNorm.py
from utils import TestCase
import torch
from torch.nn.modules.normalization import LayerNorm

class TestNNNormalizationLayerNorm(TestCase):
    def test_3d(self):
        model = LayerNorm(3)
        x = torch.randn(2, 2, 3)

        expected = model.forward(x)

        actual = self.cli(
            'LayerNorm',
            '--dim', '3',
            '--x', str(x.tolist()),
            '--param-weight', str(model.weight.tolist()),
            '--param-bias', str(model.bias.tolist()),
        )

        self.assertTensors(actual, [expected])

    def test_no_elementwise_affine(self):
        model = LayerNorm(3, elementwise_affine=False)
        x = torch.randn(2, 2, 3)

        expected = model.forward(x)

        actual = self.cli(
            'LayerNorm',
            '--dim', '3',
            '--elementwise_affine', 'false',
            '--x', str(x.tolist()),
        )

        self.assertTensors(actual, [expected])

File: src/nn/modules/normalization/LayerNorm.cpp
#include "nn/modules/normalization/LayerNorm.hpp"
#include "nn/Parameter.hpp"
#include "ggml/Context.hpp"

LayerNorm::LayerNorm(int64_t dim, float eps, bool elementwise_affine, bool bias)
    : eps_(eps), elementwise_affine_(elementwise_affine), bias_(bias)
{
    if (elementwise_affine) {
        modules["weight"] = std::make_shared<Parameter>(Tensor::Shape({dim}));

        if (bias)
            modules["bias"] = std::make_shared<Parameter>(Tensor::Shape({dim}));
    }
}

Tensor LayerNorm::forward(Context& context, Tensor x) {
    x = Tensor(*context, ggml_norm(*context, *x, eps_), x.shape());

    if (elementwise_affine_) {
        auto weight = std::static_pointer_cast<Parameter>(modules["weight"]);

        x = x * weight->forward();

        if (bias_) {
            auto bias = std::static_pointer_cast<Parameter>(modules["bias"]);

            x = x + bias->forward();
        }
    }

    return x;
}
